Replace old breadcrumb implementation in the top bar with new breadcrumb component.

MichaelMauderer opened the pull request:
### Pull Request Description

Implements #7312:  Replace the legacy code used for the project breadcrumbs and update them to the new design. 

![Peek 2023-08-21 13-40](https://github.com/enso-org/enso/assets/1428930/32d8f066-c80d-4915-8133-6dbc1edf70b3)

![Peek 2023-08-21 13-41](https://github.com/enso-org/enso/assets/1428930/a1e5aa7e-8776-4438-9c1e-80ac1afd0292)

### Important Notes

There is some ugly code around updating the theme. This should be fixed by providing a better mechanism to re-use components with different styles. The current mechanism assumes that a component will use the same theme for every instance. Implementing this is out of scope for this task, though. 

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

MichaelMauderer committed:
Replace old breadcrumb implementation with new breadcrumb component.

vitvakatu commented on app/gui/src/presenter/graph/call_stack.rs:
```diff
@@ -19,21 +20,178 @@ use ide_view as view;
+        self.view.top_bar().breadcrumbs.set_entries_from((stack, breadcrumb_index));
+    }
+
+    /// Remove the given number of levels from the stack. Will move the selection in the breadcrumbs
+    /// to the left.
+    fn shift_breadcrumb_selection_left(&self, count: usize) {
```
The comment says we remove items from the stack, but I don't see it in the implementation. Is the comment correct?

MichaelMauderer replied:
Clarified the comment. Modifying the selected item, will in turn trigger a change in the current active stack item.

vitvakatu commented on app/gui/src/presenter/graph/call_stack.rs:
```diff
@@ -19,21 +20,178 @@ use ide_view as view;
+    // needs to be taken to avoid infinite loops and any updates to the call stack should go
+    // through the breadcrumbs.
+
+    /// Method to call when existing a node in the graph editor. This will move the selection in
```
```suggestion
    /// Method to call when exiting a node in the graph editor. This will move the selection in
```

vitvakatu commented on app/gui/src/presenter/graph/call_stack.rs:
```diff
@@ -190,22 +293,28 @@ impl CallStack {
+        spawn_stream_handler(weak, graph_notifications, move |notification, _model| {
             info!("Received controller notification {notification:?}");
             match notification {
-                Notification::EnteredStack(stack) => model.add_breadcrumbs_in_view(stack),
-                Notification::ExitedStack(count) =>
-                    model.view.top_bar().breadcrumbs.pop_breadcrumbs(count),
+                Notification::EnteredStack(stack) => {
+                    debug!("Notification::EnteredStack: {:#?}", stack);
+                }
+                Notification::ExitedStack(count) => {
+                    debug!("Notification::ExitedStack: {:#?}", count);
+                }
                 _ => {}
             }
```
Do we need this handler now?

MichaelMauderer replied:
Not really. Removed it.

vitvakatu commented on lib/rust/ensogl/component/breadcrumbs/src/lib.rs:
```diff
@@ -496,10 +522,31 @@ ensogl_core::define_endpoints_2! {
         /// Currently selected breadcrumb.
-        selected(BreadcrumbId)
+        selected(BreadcrumbId),
+        /// Only emitted when the selection is changed through user interaction.
+        user_selected(BreadcrumbId),
```
Firstly, I think it's better to say that it's emitted when the selection is changed by clicking on the breadcrumb. After all, entering/leaving nodes is also a `user interaction`, but I assume it will not be emitted then.
Secondly, I don't see any usage of it in the code.

MichaelMauderer replied:
This was actually not implemented/needed in the end, so it is removed now.

vitvakatu reviewed (approved):
Only minor nitpicks

MichaelMauderer committed:
Remove unnecessary controller notifications handler in call_stack.

MichaelMauderer committed:
The function shift_breadcrumb_selection_left was updated to accurately describe its functionality. The updated comments clarify that the controller will exit the stack by the desired number of levels.

MichaelMauderer committed:
Fix typo in call_stack.rs comments.

MichaelMauderer committed:
The 'user_selected' event emitter from the Breadcrumbs component was removed. It was never used.

MichaelMauderer committed:
Merge branch 'develop' into wip/MichaelMauderer/update_project_breadcrumbs_to_reuse_cbs_ones

MichaelMauderer committed:
Merge branch 'develop' into wip/MichaelMauderer/update_project_breadcrumbs_to_reuse_cbs_ones

MichaelMauderer committed:
The commit introduces changes in the code to ensure breadcrumb history is consistent with the call stack. The internal stack history is now updated alongside the view update, and a debug assertion is added to check the equality of entries in the stack history and the visible breadcrumb.

kazcw commented:
# QA :red_circle:

1. Breadcrumb icons are offset vertically (I have checked this is not present on `develop`):
![image](https://github.com/enso-org/enso/assets/1047859/a41a91df-02c7-4d65-bb94-ea060c19ad4e)

1. Any node double-clicked is added to the breadcrumb stack, even if it isn't collapsed.
![image](https://github.com/enso-org/enso/assets/1047859/f7b21c66-0518-4dca-97dd-408fb898edc2)

1. When the breadcrumb stack contains multiple levels, their layout is intermittently incorrect (this should read `main > func4 > func5 > func6`):
![image](https://github.com/enso-org/enso/assets/1047859/8ee25d36-1732-4a42-ad07-501d476bb251)

1. Warnings like this are logged frequently:
![image](https://github.com/enso-org/enso/assets/1047859/abd496cb-8d15-4ba9-966b-094ad85fce5a)
Note the message size.

vitvakatu commented:
3 seems to be related to #7605. Let's fix it as part of this PR.

MichaelMauderer committed:
Fix icon location in breadcrumbs component.

MichaelMauderer committed:
Avoid invalid breadcrumb state after entering node that cannot be entered.

MichaelMauderer committed:
Changed warning to contain more relevant and easier to read information.

MichaelMauderer commented:
>     1. Breadcrumb icons are offset vertically (I have checked this is not present on `develop`):
>        ![image](https://user-images.githubusercontent.com/1047859/262876509-a41a91df-02c7-4d65-bb94-ea060c19ad4e.png)

Fixed
![image](https://github.com/enso-org/enso/assets/1428930/b80213a2-dc6a-4980-805f-d231715a271a)



>     2. Any node double-clicked is added to the breadcrumb stack, even if it isn't collapsed.
>        ![image](https://user-images.githubusercontent.com/1047859/262880273-f7b21c66-0518-4dca-97dd-408fb898edc2.png)

Fixed
![Peek 2023-08-24 14-17](https://github.com/enso-org/enso/assets/1428930/168c408c-fa0d-4f82-8fda-9d0134265f0c)



>     3. When the breadcrumb stack contains multiple levels, their layout is intermittently incorrect (this should read `main > func4 > func5 > func6`):
>        ![image](https://user-images.githubusercontent.com/1047859/262879806-8ee25d36-1732-4a42-ad07-501d476bb251.png)

This is not fixed yet. After internal discussion it will not be fixed with this PR but as separate work item #7605.

>     4. Warnings like this are logged frequently:
>        ![image](https://user-images.githubusercontent.com/1047859/262869112-abd496cb-8d15-4ba9-966b-094ad85fce5a.png)
>        Note the message size.

This is also fixed by changing the logging.


@kazcw It would be great if you could re-test the PR with the fixes.

kazcw commented:
# QA: :green_circle: 

The issues have been addressed. Looks good!
